### ISSUE TITLE
PortMidi Build error with MinGW64/MSYS2

### ISSUE THREAD

**Issue by sebshader** [pure-data/pure-data#2302]
Suddenly I am unable to build pd with the default configuration in Mingw
Some of this might have to do with some configuration or newer mingw packages because the same thing happens with older checkouts that previously compiled ok

gcc: /mingw64/bin/gcc: gcc.exe (Rev2, Built by MSYS2 project) 14.1.0
ld: /mingw64/bin/ld: GNU ld (GNU Binutils) 2.42
libtool: /usr/bin/libtool: libtool (GNU libtool) 2.4.7

first, there's a libtool error during libtoolize step:
```
libtoolize:   error: $pkgauxdir is not a directory: '/usr/share/libtool/build-aux'
```
(/usr/share/libtool/build-aux does exist in the mingw environment..)

using the old scripts (no ./autogen.sh) the error is
```
gcc -DPACKAGE_NAME=\"pd\" -DPACKAGE_TARNAME=\"pd\" -DPACKAGE_VERSION=\"0.55.0\" -DPACKAGE_STRING=\"pd\ 0.55.0\" -DPACKAGE_BUGREPORT=\"\" -DPACKAGE_URL=\"\" -DPACKAGE=\"pd\" -DVERSION=\"0.55.0\" -DHAVE_STDIO_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_STRINGS_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_UNISTD_H=1 -DSTDC_HEADERS=1 -DLT_OBJDIR=\".libs/\" -DHAVE_FCNTL_H=1 -DHAVE_ALLOCA=1 -Dvfork=fork -DHAVE_MALLOC=1 -DHAVE_REALLOC=1 -DRETSIGTYPE=void -DHAVE_DUP2=1 -DSTUPID_SORT=1 -I.  -I../portmidi/portmidi/pm_common -I../portmidi/portmidi/porttime -DNDEBUG   -I../portmidi/portmidi/pm_win  -DNEWBUFFER -ffast-math -fno-finite-math-only -funroll-loops -fomit-frame-pointer -O3   -MT portmidi/pm_win/pmwinmm.o -MD -MP -MF $depbase.Tpo -c -o portmidi/pm_win/pmwinmm.o portmidi/pm_win/pmwinmm.c &&\
mv -f $depbase.Tpo $depbase.Po
portmidi/pm_win/pmwinmm.c: In function 'pm_winmm_general_inputs':
portmidi/pm_win/pmwinmm.c:169:27: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
  169 |                           (void *) i, &pm_winmm_in_dictionary);
      |                           ^
portmidi/pm_win/pmwinmm.c: In function 'pm_winmm_general_outputs':
portmidi/pm_win/pmwinmm.c:209:27: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
  209 |                           (void *) i, &pm_winmm_out_dictionary);
      |                           ^
portmidi/pm_win/pmwinmm.c: In function 'winmm_in_open':
portmidi/pm_win/pmwinmm.c:471:16: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
  471 |     dwDevice = (DWORD) descriptors[i].descriptor;
      |                ^
portmidi/pm_win/pmwinmm.c: In function 'winmm_out_open':
portmidi/pm_win/pmwinmm.c:743:16: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
  743 |     dwDevice = (DWORD) descriptors[i].descriptor;
      |                ^
portmidi/pm_win/pmwinmm.c: In function 'winmm_write_byte':
portmidi/pm_win/pmwinmm.c:1046:31: error: assignment to 'uint32_t *' {aka 'unsigned int *'} from incompatible pointer type 'DWORD *' {aka 'long unsigned int *'} [-Wincompatible-pointer-types]
 1046 |         midi->fill_offset_ptr = &(hdr->dwBytesRecorded);
      |                               ^
make[2]: *** [Makefile:573: portmidi/pm_win/pmwinmm.o] Error 1
make[2]: Leaving directory '/c/Users/sebfu/codins/pd/pure-data/portmidi'
make[1]: *** [Makefile:539: all-recursive] Error 1
make[1]: Leaving directory '/c/Users/sebfu/codins/pd/pure-data'
make: *** [Makefile:442: all] Error 2
```

configure output:
```
pd 0.55.0 is now configured

    Platform:             MinGW
    Float size:           default
    Debug build:          no
    Universal build:      no
    Localizations:        yes
    Source directory:     .
    Installation prefix:  /mingw64

    Compiler:             gcc
    CPPFLAGS:             -DNDEBUG
    CFLAGS:               -ffast-math -fno-finite-math-only -funroll-loops -fomit-frame-pointer -O3
    LDFLAGS:
    INCLUDES:
    LIBS:

    External extension:   dll
    External CFLAGS:      -mms-bitfields
    External LDFLAGS:     -Wl,--enable-auto-import -no-undefined -lpd
    Deken identifier:     windows-amd64

    fftw:                 no
    wish(tcl/tk):         wish8.6
    watchdog:             no
    audio APIs:           PortAudio WASAPI MMIO
    midi APIs:            PortMidi
    libpd:                no
```

if I add ```-Wno-incompatible-pointer-types``` to CFLAGS then it gets to the linking step and freezes for hours (even SIGTERM from shell doesn't work):
```
g++ -shared \
  \
-o pd.dll \
d_arithmetic.o d_array.o d_ctl.o d_dac.o d_delay.o d_fft.o d_filter.o d_global.o d_math.o d_misc.o d_osc.o d_resample.o d_soundfile.o d_soundfile_aiff.o d_soundfile_caf.o d_soundfile_next.o d_soundfile_wave.o d_ugen.o g_all_guis.o g_array.o g_bang.o g_canvas.o g_clone.o g_editor.o g_editor_extras.o g_graph.o g_guiconnect.o g_io.o g_mycanvas.o g_numbox.o g_radio.o g_readwrite.o g_rtext.o g_scalar.o g_slider.o g_template.o g_text.o g_toggle.o g_traversal.o g_undo.o g_vumeter.o m_atom.o m_binbuf.o m_class.o m_conf.o m_glob.o m_memory.o m_obj.o m_pd.o m_sched.o s_audio.o s_inter.o s_inter_gui.o s_loader.o s_main.o s_net.o s_path.o s_print.o s_utf8.o x_acoustics.o x_arithmetic.o x_array.o x_connective.o x_file.o x_gui.o x_interface.o x_list.o x_midi.o x_misc.o x_net.o x_scalar.o x_text.o x_time.o x_vexp.o x_vexp_fun.o x_vexp_if.o d_fft_fftsg.o s_file.o s_midi.o s_audio_mmio.o s_audio_pa.o s_midi_pm.o s_audio_paring.o \
-lm    ../portaudio/libportaudio.a   ../portmidi/libportmidi.a  \
    -lws2_32 -lwinmm -lole32 -static-libgcc -static-libstdc++ \
-Wl,--export-all-symbols \
    -Wl,--out-implib=pd.lib -Wl,--output-def=pd.def
/bin/sh ../libtool  --tag=CC   --mode=link gcc -DPD -DPD_INTERNAL -DINSTALL_PREFIX=\"/mingw64\" -ffast-math -fno-finite-math-only -funroll-loops -fomit-frame-pointer -O3   -DWISH='"wish8.6"'   -DEXTERNAL_EXTENSION="'.dll'"    -DUSEAPI_MMIO -DUSEAPI_PORTAUDIO    -DWINVER=0x0600 -D_WIN32_WINNT=0x0600 -Wno-incompatible-pointer-types -mwindows   -o pd.exe pd-s_entry.o  pd.res pd.lib      -lws2_32 -lwinmm -lole32 -static-libgcc -static-libstdc++
```

6e3318555a32ab1bb1b4e6a08136dabf675803f1 built ok last month & now doesn't..

**Comment by Lucarda** [comment]
can confirm

```
portmidi/pm_win/pmwinmm.c: In function 'winmm_write_byte':
portmidi/pm_win/pmwinmm.c:1046:31: error: assignment to 'uint32_t *' {aka 'unsigned int *'} from incompatible pointer type 'DWORD *' {aka 'long unsigned int *'} [-Wincompatible-pointer-types]
 1046 |         midi->fill_offset_ptr = &(hdr->dwBytesRecorded);
      |  
```

using the latest **Msys2** and following instructions from 

https://puredata.info/docs/developer/WindowsMinGW

**Comment by sebshader** [comment]
hmm I think the libtoolize issue might be because I had C:\Program Files\Git on the path before, which has its own mingw64 directory

**Comment by sebshader** [comment]
ok, so all the issues besides the one @Lucarda confirmed were due to paths added by me. Sorry

**Comment by Lucarda** [comment]
> Some of this might have to do with some configuration or newer mingw packages because

@sebshader  you are right


|compiles fine | compilation fails|
|----------|----------|
|gcc version 13.2.0 |gcc.exe (Rev2, Built by MSYS2 project) 14.1.0|
|GNU ld (GNU Binutils) 2.42  |GNU ld (GNU Binutils) 2.42|
|libtool (GNU libtool) 2.4.7 |libtool (GNU libtool) 2.4.7|


we should add that `-Wno-incompatible-pointer-types` flag.

**Comment by sebshader** [comment]
Upgrading to the latest portmidi fixes it (which btw is in git, not in subversion like the update.sh script uses afaict)

**Comment by Spacechild1** [comment]
Seems like GCC 14 has turned some warnings to errors by default. See https://news.ycombinator.com/item?id=40284408.

> we should add that -Wno-incompatible-pointer-types flag

The solution is not to disable the warning, but rather not make it an error, i.e. `-Wno-error=incompatible-pointer-types`.

We *could* update portmidi, but it's probably not a good idea to do this right before a release. I'd suggest to do that at a later point.

**Comment by umlaeute** [comment]
yep.
recent compilers are getting stricter policies, which lead to FTBFS errors (see e.g. 47ce79fadbf6909382656818f2b7d09236f97a62)

we *should* strive to fix these problems (rather than hide them explicitly), and this most likely involves updating PortMidi - but the timing is bad.

**Comment by umlaeute** [comment]
afaict this has been fixed upstream in https://github.com/PortMidi/portmidi/commit/dba83573fe43409fa259cf3c9583f8446085e5e5

as of writing, there hasn't been a new release of PortMidi yet, that contains this fix

**Comment by umlaeute** [comment]
i noticed that miller hand-applied a fix for this in b00725132b13a29566b9188f01a6bf277c5c7a07, thus closing (we should still update portaudio and portmidi)
